smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/ServiceValidator.java

```diff
@@ -16,55 +16,210 @@
-                topDownIndex.getContainedResources(shape.getId()));
-        if (!duplicateResourceNames.isEmpty()) {
-            events.add(conflictingNames(shape, "resources", duplicateResourceNames));
+        if (conflicts.isEmpty()) {
+            return Collections.emptyList();
         }
 
-        // Ensure that operations bound to the service have unique shape names.
-        Map<String, List<ShapeId>> duplicateOperationNames = ValidationUtils.findDuplicateShapeNames(
-                topDownIndex.getContainedOperations(shape));
-        if (!duplicateOperationNames.isEmpty()) {
-            events.add(conflictingNames(shape, "operations", duplicateOperationNames));
+        // Determine the severity of each conflict.
+        ConflictDetector detector = new ConflictDetector(model);
+        List<ValidationEvent> events = new ArrayList<>();
+
+        // Figure out if each conflict can be ignored, and then emit events for
+        // both sides of the conflict using the appropriate severity.
+        for (Map.Entry<String, List<ShapeId>> entry : conflicts.entrySet()) {
+            List<ShapeId> ids = entry.getValue();
+            for (int i = 0; i < ids.size(); i++) {
+                Shape subject = model.expectShape(ids.get(i));
+                for (int j = 0; j < ids.size(); j++) {
+                    if (i != j) {
+                        Shape other = model.expectShape(ids.get(j));
+                        Severity severity = detector.detect(subject, other);
+                        if (severity != null) {
+                            events.add(conflictingNames(severity, service, subject, other));
+                            events.add(conflictingNames(severity, service, other, subject));
+                        }
+                    }
+                }
+            }
         }
 
         return events;
     }
 
-    private ValidationEvent conflictingNames(ServiceShape shape, String descriptor, Map<String, List<ShapeId>> dupes) {
-        return error(shape, String.format(
-                "All %s contained within a service hierarchy must have case-insensitively unique names regardless of "
-                + "their namespaces. The following %s were found in this service to have conflicting names: %s",
-                descriptor, descriptor, dupes));
+    private ValidationEvent conflictingNames(Severity severity, ServiceShape shape, Shape subject, Shape other) {
+        // Whether it's a should or a must based on the severity.
+        String declaration = severity == Severity.DANGER || severity == Severity.ERROR ? "must" : "should";
+
+        return ValidationEvent.builder()
+                .eventId(getName())
+                .severity(severity)
+                .shape(subject)
+                .message(String.format(
+                        "Shape name `%s` conflicts with `%s` in the `%s` service closure. Shapes in the closure "
```

Comment:
Subtle tweak to the wording - "These shapes in the closure..." - to point that the sentence is referring to explicitly these shapes. I can foresee some reading the last part and being confused when encountering this taken from the test outputs, with one saying must and the other saying should worded to a general case:

```
[DANGER] another.ns#ListC: Shape name `ListC` conflicts with `ns.foo#ListC` in the `ns.foo#MyService` service closure. Shapes in the closure of a service must have case-insensitively unique names regardless of their namespaces. | Service
[NOTE] ns.foo#String: Shape name `String` conflicts with `smithy.api#String` in the `ns.foo#MyService` service closure. Shapes in the closure of a service should have case-insensitively unique names regardless of their namespaces. | Service
```